### PR TITLE
Fix floating point precision

### PR DESCRIPTION
**Pull request by OrigamiDream**
## :recycle: Current situation

Related:
- #955 
- #956 

## :bulb: Proposed solution

As @adriancable have mentioned in #956, the previous solution may have possible incorrect behaviors.
So I have tried all possible tests which can be happened in 0 and 100.

I moved the rounding logic in front of clamping logic.
Plus, to fix in case the error is smaller than epsilon and greater than 0.5, I have changed `Math.round()` to `Math.floor()`. 
We don't need round the value in server-side, due to Apple Home app will always automatically round the value fit to step size.

### Testing

I have rewritten entire test code related with this issue.
The `minStep` in the final test only uses (100/6), of course I have tried other combinations like 3, 7, 9.
All cases have passed, including approaches which @adriancable have mentioned before.

The test takes up to 3 minutes per step, this is the reason why I have decided to use only 100/6.

### Reviewer Nudging

I would appreciate if you have better idea.

### CODE REVIEW

**Comment by Supereg** [comment]
@adriancable would you mind reviewing this PR?

**Comment by OrigamiDream** [comment]
The failure is originated from the new rounding algorithm. Im gonna fix this after WWDC 🥳

**Comment by adriancable** [comment]
It’s not sufficient to check that stepValue != null. It could also be undefined or 0. Either will lead to an overflow which is what happens in your test.

**Comment by OrigamiDream** [comment]
We should not handle stepValue which is undefined or 0.
The original input value must not be changed in this case.
Instead, we can just warn to the developers.

And the test failure is not a kind of overflow.
They are just the wrong 'numericMin' handling errors in the old test cases.

**Comment by adriancable** [comment]
I think this looks OK now.

File: src/lib/Characteristic.spec.ts [hunk 1]
```diff
@@ -805,6 +805,7 @@ describe("Characteristic", () => {
       const characteristic = createCharacteristicWithProps({
         format: Formats.FLOAT,
         minStep: 0.001,
+        minValue: 0,
```
Comment (Supereg) [inline src/lib/Characteristic.spec.ts:808]: Was this change required to make the test work again or what’s the reason behind it?

**Review by Supereg — COMMENTED** [review]
Looks good to me. There were some other edge cases discussed in the issue form right? Would it make sense to add them to the suite of test cases to avoid any future regressions?

File: src/lib/Characteristic.spec.ts (hunk 1 above)
Comment (OrigamiDream) [reply src/lib/Characteristic.spec.ts:808]: The `minValue` changes to `-Number.MAX_VALUE` internally, which have many decimal points.
Old rounding algorithm does not care about the decimal points, that the reason why the test case had worked.
Since it was working wrongly, I have updated the test case to fit with the new rounding algorithm.

Old algorithm:
- 1.024 -> 1.024 (because the `minValue` is treated as no decimal points)

New algorithm:
- 1.024 + (decimal points of `-Number.MAX_VALUE`) -> 1.7976931348623157**e+308**

To the issue be resolved, the old test case must be changed.

**Comment by OrigamiDream** [comment]
Possibly one edge case.

Basically all characteristics have `minValue` itself, so there would no failure if `minValue` is valid.
However somehow the plugin developer set the `minValue` as `undefined` ignoring default values, the `minValue` will be changed to `-Number.MAX_VALUE`.
Eventually, the rounding algorithm would work wrongly. (It would feel like unexpected behavior to the plugin developer)

Although this can be happened, it is more like the developer sets non-null properties as null which cannot be detected in compile-time.
Therefore, I have decided to not add warning signs about this. (And too many affected test cases which are spying on `characteristicWarning`)

**Review by Supereg — APPROVED** [review]
